sky/backends/cloud_vm_ray_backend.py

```diff
@@ -1611,64 +1586,62 @@ def _retry_zones(
+                        cluster_yaml=handle.cluster_yaml,
+                        is_prev_cluster_healthy=is_prev_cluster_healthy,
+                        log_dir=self.log_dir)
+                    # NOTE: We will handle the logic of '_ensure_cluster_ray_started'
+                    # in 'provision_utils.post_provision_runtime_setup()' in the caller.
                     resources_vars = (
                         to_provision.cloud.make_deploy_resources_variables(
                             to_provision, handle.cluster_name_on_cloud, region,
                             zones))
                     config_dict['provision_record'] = provision_record
                     config_dict['resources_vars'] = resources_vars
                     config_dict['handle'] = handle
-                    return config_dict
-
-                # NOTE: We try to cleanup the cluster even if the previous
-                # cluster does not exist. Also we are fast at
-                # cleaning up clusters now if there is no existing node..
-                CloudVmRayBackend().post_teardown_cleanup(
-                    handle, terminate=not is_prev_cluster_healthy)
-                # TODO(suquark): other clouds may have different zone
-                #  blocking strategy. See '_update_blocklist_on_error'
-                #  for details.
-                if zones is None:
-                    self._blocked_resources.add(to_provision.copy(zone=None))
+                    tpu_name = config_dict.get('tpu_name')
+                    if tpu_name is None:
+                        return config_dict
+                    # tpu_name will only be set when TPU node (not TPU VM)
+                    # is required.
+                    logger.info(
+                        f'{colorama.Style.BRIGHT}Provisioning TPU node on '
+                        f'{to_provision.cloud} '
+                        f'{region.name}{colorama.Style.RESET_ALL}{zone_str}')
+
+                    success = self._try_provision_tpu(to_provision, config_dict)
+                    if success:
+                        return config_dict
+                    raise RuntimeError('Failed to provision TPU node.')
+                except Exception as e:  # pylint: disable=broad-except
+                    # NOTE: We try to cleanup the cluster even if the previous
+                    # cluster does not exist. Also we are fast at
+                    # cleaning up clusters now if there is no existing node..
+                    CloudVmRayBackend().post_teardown_cleanup(
+                        handle, terminate=not is_prev_cluster_healthy)
+                    # TODO(suquark): other clouds may have different zone
+                    #  blocking strategy. See '_update_blocklist_on_error'
+                    #  for details.
+                    FailoverCloudErrorHandlerV2.update_blocklist_on_error(
+                        self._blocked_resources, to_provision, region, zones, e)
```

Comment:
Q: where do we failover from V2 (handler not found) -> V1?

Reply:
We don't failover to V1 now. For the cloud using the new provisioner API, it should use V2 only; and for the cloud using the old ray autoscaler, it should use V1 only.
Previously, we manually check if the TPU VM is used to decide whether to go through the new provisioner code path or the ray autoscaler path. Now, all GCP resources are using new provisioner API, so we use V2 only for it.

Reply:
Ah I see, after carefully looking into the function this makes sense. `_retry_zones()` is ~350 LoC, we should refactor it to make it easier to read.

Reply:
Let's leave the refactoring in another PR for the simplicity of this PR. ; )